README: add enclave-cc introduction and motivation (#41)
confidential-containers/enclave-cc: README.md

# enclave-cc
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fconfidential-containers%2Fenclave-cc.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2Fconfidential-containers%2Fenclave-cc?ref=badge_shield)
## Introduction

>   Confidential containers are the product of the combination of confidential computing technology and cloud-native technology. Confidential containers are hardware-based (HW-TEE) for resource isolation, data encryption, and remote attestation, and can prevent theft and tampering of data in the execution environment by cloud service providers and any high-privileged third parties, thus effectively securing user data and assets.

Reviewer: ```suggestion
  Confidential containers are the product of the combination of confidential computing technology and cloud-native technology. Confidential containers use hardware-based Trusted Execution Environments (HW-TEE) for resource isolation, data encryption, and remote attestation. They can protect data in execution from cloud service providers and other high-privileged third parties.
```

Reviewer 2: some of wording changes IMO:
data encryption -> data protection
high-privileged -> privileged

>   Confidential containers are the product of the combination of confidential computing technology and cloud-native technology. Confidential containers are hardware-based (HW-TEE) for resource isolation, data encryption, and remote attestation, and can prevent theft and tampering of data in the execution environment by cloud service providers and any high-privileged third parties, thus effectively securing user data and assets.

Reviewer 2: some of wording changes IMO:
- data encryption -> data protection
- high-privileged -> privileged

>   Confidential containers have the following characteristics: they provide confidentiality, integrity and security protection for data, especially for runtime data; they provide full-link security for the deployment and operation of sensitive workloads; they provide a way to prove that the environment in which confidential workloads are currently running is real and trustworthy; tenants maintain almost the same experience as using ordinary containers, and deploy sensitive applications without operation and maintenance costs.

Author: > 

Thanks Dan, About "without operation and maintenance costs" should be "without additional O&M costs". But we can remove this sentence.

>   Confidential containers have the following characteristics: they provide confidentiality, integrity and security protection for data, especially for runtime data; they provide full-link security for the deployment and operation of sensitive workloads; they provide a way to prove that the environment in which confidential workloads are currently running is real and trustworthy; tenants maintain almost the same experience as using ordinary containers, and deploy sensitive applications without operation and maintenance costs.

Reviewer: oh sure.. "without additional..." make sense too. either way you prefer.

>   The Confidential Containers project provides a way to protect cloud native applications in HW-TEE without additioanl modification to the container image during deployment. HW-TEE-based confidential container isolation can take two forms: process and virtual machine based isolation.

Reviewer: ```suggestion
  The Confidential Containers project provides a way to protect cloud native applications in a HW-TEE without additional modification to the container image during development. HW-TEE-based confidential container isolation can take two forms: process and virtual machine based isolation.
```

>   Enclave-cc provides process-based confidential container solution leveraging Intel SGX. The term enclave is often associated with process-based isolation. 

Reviewer 3: please change `Enclave-cc provides` to `Enclave-cc provides a`

## Motivation

  Enclave-CC enables process-based isolation. Process-based isolation is beneficial by drawing the isolation boundary exactly around each container process. This reduces the Trusted Computing Base, or the amount of the system your application depends on for security.

>   Consistent with the VM-based isolation design, the enclave-cc approach does not measure and attest to the workload (user application). Instead it uses generic enclaves to isolate user's application from the rest of the system. Measurement and attestation of the empty generic enclaves is used in building trust to deliver the application securely.

Reviewer 3: please change `is` in `Measurement and attestation of the empty generic enclaves is` to `are `

>   Enclave-CC also follows up the trust model of confidential containers, provides a complete security solution, from creating user application image with encryption/signature to uploading it to image registry in user side, to pulling the image, verifying its signature and decrypting it, unpacking it into user rootfs in enclave, then booting up user containers and mounting encrypted rootfs, all the processes are guaranteed to be secure.

Reviewer: ```suggestion
  The Confidential Containers project, including the Enclave-CC approach provides all the elements for a complete deployment flow, from creating user container image with encryption/signature to uploading it to an image registry, to pulling the image, verifying its signature and decrypting it, unpacking it into user rootfs in enclave, then booting up user containers and mounting encrypted rootfs, all the container processes are isolated with enclaves.
```

they provide a way to prove that the environment in which confidential workloads are launched is authentic and trustworthy;
tenants maintain almost the same experience as using ordinary containers, but can deploy sensitive applications with greater
confidence in their security.

The Confidential Containers project provides a way to protect cloud native applications in a HW-TEE without additional modification
to the container image during development. HW-TEE-based confidential container isolation can take two forms: process and virtual
machine based isolation.

Enclave-cc provides a process-based confidential container solution leveraging Intel SGX. The term enclave is often associated with
process-based isolation.

## Motivation

Enclave-CC enables process-based isolation. Process-based isolation is beneficial by drawing the isolation boundary exactly around
each container process. This reduces the Trusted Computing Base, or the amount of the system your application depends on for security.

Consistent with the VM-based isolation design, the enclave-cc approach does not measure and attest to the workload (user application).

> Instead it uses generic enclaves to isolate user's application from the rest of the system. Measurement and attestation of the empty

Reviewer 4: Since this sentence refers to "app enclave" only, it's probably better to say:

```suggestion
Instead it uses a generic enclave to isolate user's application from the rest of the system. Measurement and attestation of the empty
```
